pkg/virt-api/webhooks/hyperv.go

```diff
@@ -178,16 +177,27 @@ func ValidateVirtualMachineInstanceHypervFeatureDependencies(field *k8sfield.Pat
 		}
 	}
 
-	if spec.Domain.Features != nil && spec.Domain.Features.Hyperv != nil && spec.Domain.Features.Hyperv.EVMCS != nil {
-		if spec.Domain.CPU == nil || spec.Domain.CPU.Features == nil || len(spec.Domain.CPU.Features) == 0 {
-			causes = append(causes, metav1.StatusCause{Type: metav1.CauseTypeFieldValueRequired, Message: "vmx cpu feature is required when evmcs is set", Field: "spec.domain.cpu.features"})
-		} else if spec.Domain.CPU != nil || spec.Domain.CPU.Features != nil {
-			for i, f := range spec.Domain.CPU.Features {
-				if f.Name == nodelabellerutil.VmxFeature {
-					if f.Policy != nodelabellerutil.RequirePolicy {
-						causes = append(causes, metav1.StatusCause{Type: metav1.CauseTypeFieldValueInvalid, Message: "vmx cpu feature has to be set to " + nodelabellerutil.RequirePolicy + " policy", Field: "spec.domain.cpu.features[" + strconv.Itoa(i) + "].policy"})
-					}
-				}
+	if spec.Domain.Features == nil || spec.Domain.Features.Hyperv == nil || spec.Domain.Features.Hyperv.EVMCS == nil {
```

Comment:
No, we can't :)

From this point on, all of the checks are regarding EVMCS. In the upper check, we general HyperV dependencies, not only EVMCS ones.

You can look at the original condition and see that no logic is changed. This commit is 100% a refactoring that has no effect on actual behavior.

Reply:
What I meant by this is: Why should we actually call `getHypervFeatureDependencies` if `spec.Domain.Features == nil || spec.Domain.Features.Hyperv == nil`. But I see it would just return immediately. Moving this to the top would make it more obvious imo.

Reply:
> But I see it would just return immediately

I don't see a point in checking this if it's already being checked in `getHypervFeatureDependencies()`. I call the function, it will tell me if there are dependencies or not. I don't need to even know what it's checking inside and I see no reason to explicitly check it.

Reply:
But these are different checks..
Currently we:
- call getHypervFeatureDependencies()
  - Which inside checks if `spec.Domain.Features == nil || spec.Domain.Features.Hyperv == nil` and if so it returns
- check if EVMCS is enabled (by checking all 3 conditions, e.g. `Features == nil || Features.Hyperv != nil || Features.Hyperv.EVMCS == nil`

The check for EVMCS cannot be done at the top since even if it's not enabled we should take care of getHypervFeatureDependencies().

We can instead:
- Check for hypev features
- Call getHypervFeatureDependencies() (which, again, would do the same check again internally)
- Check for EVMCS (with exactly the same conditions as now).

I don't think it's more obvious, quite the contrary. I want to check for dependencies, and get an empty feature slice if there are none. I don't care how the function checks it and I don't need to re-check the same condition it checks. Another point is that if someday these conditions are changed we now have to change them in two different places. I honestly don't think it's any better.

But we can agree to disagree and move on.

Reply:
We're trying to optimize a single if statement ... ;)
If we really want to optimize it, let's do it in a follow-up. 

Btw, `getHypervFeatureDependencies` can't even return nil.